README.md
# Php Reserved Words checker

[![Latest Version](https://img.shields.io/github/v/release/sspat/reserved-words)](https://github.com/sspat/reserved-words/releases)
[![Build](https://img.shields.io/travis/sspat/reserved-words/master)](https://travis-ci.org/sspat/reserved-words)
[![Mutation testing badge](https://img.shields.io/endpoint?style=flat&url=https%3A%2F%2Fbadge-api.stryker-mutator.io%2Fgithub.com%2Fsspat%2Freserved-words%2Fmaster)](https://dashboard.stryker-mutator.io/reports/github.com/sspat/reserved-words/master)
[![Test Coverage](https://coveralls.io/repos/github/sspat/reserved-words/badge.svg?branch=master)](https://coveralls.io/github/sspat/reserved-words?branch=master)
[![Type Coverage](https://shepherd.dev/github/sspat/reserved-words/coverage.svg)](https://shepherd.dev/github/sspat/reserved-words)
[![License](https://img.shields.io/github/license/sspat/reserved-words)](https://github.com/sspat/reserved-words/blob/master/LICENSE)

## About

This package allows checking strings for being a PHP reserved word and also the possibility of using
the string as a PHP namespace/class/interface/trait, function, method or constant name.

By default checks are performed for your current runtime PHP version, but you can specify any version to check against.

This can come in handy during code generation for example, when PHP code is generated based on some user input.

Reserved words in the PHP documentation:
- [List of Keywords](https://www.php.net/manual/en/reserved.keywords.php)
- [List of other reserved words](https://www.php.net/manual/en/reserved.other-reserved-words.php)


## Installation

```
composer require sspat/reserved-words
```

## Usage

```php
<?php

use sspat\ReservedWords\ReservedWords;

$reservedWords = new ReservedWords();
$word = 'list';

/**
 * Checks that the word is reserved in any PHP version.
 * It is still possible to use reserved words in php code in many places, but generally you should avoid it.
 * This method also returns true for words that are marked as "soft" reserved in the PHP docs and may
 * become reserved in future versions of the language.
 */
$isReserved = $reservedWords->isReserved($word);
/**
 * Checks that the word cannot be used as a constant name in your current php version.
 */
$cannotUseAsConstantName = $reservedWords->isReservedConstantName($word);
/**
 * Checks that the word cannot be used as a namespace part in your current php version.
 * 
 * This is used for checking parts of namespaces, not full namespace strings.
 * E.g. calling this with `Some\Namespace\String` is incorrect, you should make three separate calls
 * with `Some`, `Namespace` and `String`.
 */
$cannotUseAsNamespaceName = $reservedWords->isReservedNamespaceName($word);
/**
 * Checks that the word cannot be used as a class/interface/trait name in your current php version.
 */
$cannotUseAsNamespaceName = $reservedWords->isReservedClassName($word);
/**
 * Checks that the word cannot be used as a function name in your current php version.
 */
$cannotUseAsFunctionName = $reservedWords->isReservedFunctionName($word);
/**
 * Checks that the word cannot be used as a method name in your current php version.
 */
$cannotUseAsMethodName = $reservedWords->isReservedMethodName($word);

/**
 * The following methods also accept a second parameter, to check against a PHP version different than your current runtime
 */
$cannotUseAsConstantName = $reservedWords->isReservedConstantName($word, '5.6');
$cannotUseAsNamespaceName = $reservedWords->isReservedNamespaceName($word, '5.6.1');
$cannotUseAsNamespaceName = $reservedWords->isReservedClassName($word, '5.6.1');
$cannotUseAsFunctionName = $reservedWords->isReservedFunctionName($word, '8.0');
$cannotUseAsMethodName = $reservedWords->isReservedMethodName($word, '7.4.2');
```


CHANGELOG.md
### 3.0.0
* Added support for PHP 8.1 
* Added support for thecodingmachine/safe v2

### 2.0.0

* Added support for PHP 8.0
* BC Break: Dropped support of PHP 7.2
* BC Break: Added new keyword reserved in php 8 - `match`
* BC Break: Updated `mixed` keyword to reflect changes in php 8
* BC Break: Updated all keywords except `namespace` and `__halt_compiler` 
  to reflect that they can be used as namespace parts since php 8.0
* BC Break: Added new method `isReservedClassName` to check for reserved words in 
  class/trait/interface names. The method `isReservedNamespaceName` should
  be now used only for checking for reserved words in namespaces.
  These changes are necessary because php 8 changed the reserved words behavior for namespaces.
* Migrated CI to GitHub actions
* Added mutation testing
* Added phpstan static analysis


tests/ReservedWordsTest.php
<?php

declare(strict_types=1);

namespace sspat\ReservedWords\Tests;

use PHPUnit\Framework\Assert;
use PHPUnit\Framework\TestCase;
use sspat\ReservedWords\ReservedWords;
use sspat\ReservedWords\ReservedWordsLookupError;

use function strtoupper;

/**
 * @covers \sspat\ReservedWords\ReservedWords
 */
final class ReservedWordsTest extends TestCase
{
    public function testDefaultReservedWordsLoaded(): void
    {
        $reservedWord  = 'list';
        $reservedWords = new ReservedWords();

        Assert::assertTrue($reservedWords->isReserved($reservedWord));
    }

    public function testIsReserved(): void
    {
        $reservedWord  = 'reserved-word';
        $reservedWords = new ReservedWords([$reservedWord => []]);

        Assert::assertTrue($reservedWords->isReserved($reservedWord));
    }

    public function testIsNotReserved(): void
    {
        $notReservedWord = 'not-reserved-word';
        $reservedWords   = new ReservedWords([]);

        Assert::assertFalse($reservedWords->isReservedMethodName($notReservedWord));
    }

    public function testIsReservedCaseInsensitive(): void
    {
        $reservedWord  = 'reserved-word';
        $reservedWords = new ReservedWords([
            $reservedWord => [
                'constant' => '7.0',
                'namespace' => '7.0',
                'function' => '7.0',
                'method' => '7.0',
            ],
        ]);

        Assert::assertTrue($reservedWords->isReservedMethodName(strtoupper($reservedWord), '7.0'));
    }

    /**
     * @param array<string, string|bool|array<array-key, string>> $reservedWordParameters
     *
     * @dataProvider reservedWordsList
     */
    public function testConstantName(
        string $reservedWord,
        array $reservedWordParameters,
        string $phpVersion,
        bool $isReserved
    ): void {
        $reservedWords = new ReservedWords([$reservedWord => $reservedWordParameters]);

        Assert::assertEquals($isReserved, $reservedWords->isReservedConstantName($reservedWord, $phpVersion));
    }

    /**
     * @param array<string, string|bool|array<array-key, string>> $reservedWordParameters
     *
     * @dataProvider reservedWordsList
     */
    public function testNamespaceName(
        string $reservedWord,
        array $reservedWordParameters,
        string $phpVersion,
        bool $isReserved
    ): void {
        $reservedWords = new ReservedWords([$reservedWord => $reservedWordParameters]);

        Assert::assertEquals($isReserved, $reservedWords->isReservedNamespaceName($reservedWord, $phpVersion));
    }

    /**
     * @param array<string, string|bool|array<array-key, string>> $reservedWordParameters
     *
     * @dataProvider reservedWordsList
     */
    public function testClassName(
        string $reservedWord,
        array $reservedWordParameters,
        string $phpVersion,
        bool $isReserved
    ): void {
        $reservedWords = new ReservedWords([$reservedWord => $reservedWordParameters]);

        Assert::assertEquals($isReserved, $reservedWords->isReservedClassName($reservedWord, $phpVersion));
    }

    /**
     * @param array<string, string|bool|array<array-key, string>> $reservedWordParameters
     *
     * @dataProvider reservedWordsList
     */
    public function testFunctionName(
        string $reservedWord,
        array $reservedWordParameters,
        string $phpVersion,
        bool $isReserved
    ): void {
        $reservedWords = new ReservedWords([$reservedWord => $reservedWordParameters]);

        Assert::assertEquals($isReserved, $reservedWords->isReservedFunctionName($reservedWord, $phpVersion));
    }

    /**
     * @param array<string, string|bool|array<array-key, string>> $reservedWordParameters
     *
     * @dataProvider reservedWordsList
     */
    public function testMethodName(
        string $reservedWord,
        array $reservedWordParameters,
        string $phpVersion,
        bool $isReserved
    ): void {
        $reservedWords = new ReservedWords([$reservedWord => $reservedWordParameters]);

        Assert::assertEquals($isReserved, $reservedWords->isReservedMethodName($reservedWord, $phpVersion));
    }

    public function testInvalidPhpVersion(): void
    {
        $reservedWord  = 'reserved-word';
        $reservedWords = new ReservedWords([$reservedWord => []]);

        $this->expectException(ReservedWordsLookupError::class);
        $this->expectErrorMessage('Invalid PHP version: 7, the correct format is: /^\d\.\d\.?\d*$/');
        $reservedWords->isReservedMethodName($reservedWord, '7');
    }

    /**
     * @return array<array-key, array<string, string|bool|array<string, string|bool|array<array-key, string>>>>
     */
    public function reservedWordsList(): array
    {
        return [
            [
                'reservedWord' => 'reserved-word',
                'reservedWordParameters' => [
                    'constant' => false,
                    'namespace' => false,
                    'class' => false,
                    'function' => false,
                    'method' => false,
                ],
                'phpVersion' => '4.0.0',
                'isReserved' => false,
            ],
            [
                'reservedWord' => 'reserved-word',
                'reservedWordParameters' => [
                    'constant' => '7.0',
                    'namespace' => '7.0',
                    'class' => '7.0',
                    'function' => '7.0',
                    'method' => '7.0',
                ],
                'phpVersion' => '7.0',
                'isReserved' => true,
            ],
            [
                'reservedWord' => 'reserved-word',
                'reservedWordParameters' => [
                    'constant' => '7.0',
                    'namespace' => '7.0',
                    'class' => '7.0',
                    'function' => '7.0',
                    'method' => '7.0',
                ],
                'phpVersion' => '7.0.0',
                'isReserved' => true,
            ],
            [
                'reservedWord' => 'reserved-word',
                'reservedWordParameters' => [
                    'constant' => '7.0',
                    'namespace' => '7.0',
                    'class' => '7.0',
                    'function' => '7.0',
                    'method' => '7.0',
                ],
                'phpVersion' => '7.0.1',
                'isReserved' => true,
            ],
            [
                'reservedWord' => 'reserved-word',
                'reservedWordParameters' => [
                    'constant' => '7.0',
                    'namespace' => '7.0',
                    'class' => '7.0',
                    'function' => '7.0',
                    'method' => '7.0',
                ],
                'phpVersion' => '5.6',
                'isReserved' => false,
            ],
            [
                'reservedWord' => 'reserved-word',
                'reservedWordParameters' => [
                    'constant' => ['5.0', '7.0'],
                    'namespace' => ['5.0', '7.0'],
                    'class' => ['5.0', '7.0'],
                    'function' => ['5.0', '7.0'],
                    'method' => ['5.0', '7.0'],
                ],
                'phpVersion' => '4.0',
                'isReserved' => false,
            ],
            [
                'reservedWord' => 'reserved-word',
                'reservedWordParameters' => [
                    'constant' => ['5.0', '7.0'],
                    'namespace' => ['5.0', '7.0'],
                    'class' => ['5.0', '7.0'],
                    'function' => ['5.0', '7.0'],
                    'method' => ['5.0', '7.0'],
                ],
                'phpVersion' => '5.0',
                'isReserved' => true,
            ],
            [
                'reservedWord' => 'reserved-word',
                'reservedWordParameters' => [
                    'constant' => ['5.0', '7.0'],
                    'namespace' => ['5.0', '7.0'],
                    'class' => ['5.0', '7.0'],
                    'function' => ['5.0', '7.0'],
                    'method' => ['5.0', '7.0'],
                ],
                'phpVersion' => '5.4',
                'isReserved' => true,
            ],
            [
                'reservedWord' => 'reserved-word',
                'reservedWordParameters' => [
                    'constant' => ['5.0', '7.0'],
                    'namespace' => ['5.0', '7.0'],
                    'class' => ['5.0', '7.0'],
                    'function' => ['5.0', '7.0'],
                    'method' => ['5.0', '7.0'],
                ],
                'phpVersion' => '7.0',
                'isReserved' => false,
            ],
            [
                'reservedWord' => 'reserved-word',
                'reservedWordParameters' => [
                    'constant' => ['5.0', '7.0'],
                    'namespace' => ['5.0', '7.0'],
                    'class' => ['5.0', '7.0'],
                    'function' => ['5.0', '7.0'],
                    'method' => ['5.0', '7.0'],
                ],
                'phpVersion' => '7.1',
                'isReserved' => false,
            ],
        ];
    }
}


tests/ReservedWordsLookupErrorTest.php
<?php

declare(strict_types=1);

namespace sspat\ReservedWords\Tests;

use PHPUnit\Framework\TestCase;
use sspat\ReservedWords\ReservedWordsLookupError;

/**
 * @covers \sspat\ReservedWords\ReservedWordsLookupError
 */
final class ReservedWordsLookupErrorTest extends TestCase
{
    public function testInvalidPhpVersion(): void
    {
        $this->expectExceptionMessage('Invalid PHP version: 7, the correct format is: /\d+/');

        throw ReservedWordsLookupError::invalidPhpVersion('7', '/\d+/');
    }
}


src/ReservedWordsList.php
<?php

declare(strict_types=1);

namespace sspat\ReservedWords;

final class ReservedWordsList
{
    /**
     * Array keys represent the reserved words in php
     *
     * The nested array keys have the following meaning:
     * "constant" - whether the reserved word can be used as a constant name
     * "namespace" - whether the reserved word can be used as part of a namespace
     * "class" - whether the reserved word can be used as a class, interface or trait name
     * "function" - whether the reserved word can be used as a function name
     * "method" - whether the reserved word can be used as a method name
     *
     * The nested array values have the following meaning:
     * "false" - can be used in any currently released php version
     * "5.1" - cannot be used since the specified php version until the latest currently released php version
     * "[5.1, 7.0]" - cannot be used from the specified lower php version up to the upper version, but excluding the upper version itself
     */
    public const PHP_RESERVED_WORDS = [
        '__halt_compiler' => [
            'constant' => ['5.1', '7.0'],
            'namespace' => '5.1',
            'class' => '5.1',
            'function' => '5.1',
            'method' => ['5.1', '7.0'],
        ],
        'abstract' => [
            'constant' => ['5.0', '7.0'],
            'namespace' => ['5.0', '8.0'],
            'class' => '5.0',
            'function' => '5.0',
            'method' => ['5.0', '7.0'],
        ],
        'and' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'array' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'as' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'break' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'callable' => [
            'constant' => ['5.4', '7.0'],
            'namespace' => ['5.4', '8.0'],
            'class' => '5.4',
            'function' => '5.4',
            'method' => ['5.4', '7.0'],
        ],
        'case' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'catch' => [
            'constant' => ['5.0', '7.0'],
            'namespace' => ['5.0', '8.0'],
            'class' => '5.0',
            'function' => '5.0',
            'method' => ['5.0', '7.0'],
        ],
        'class' => [
            'constant' => '4.0',
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'clone' => [
            'constant' => ['5.0', '7.0'],
            'namespace' => ['5.0', '8.0'],
            'class' => '5.0',
            'function' => '5.0',
            'method' => ['5.0', '7.0'],
        ],
        'const' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'continue' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'declare' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'default' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'die' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'do' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'echo' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'else' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'elseif' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'empty' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'enddeclare' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'endfor' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'endforeach' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'endif' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'endswitch' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'endwhile' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'eval' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'exit' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'extends' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'final' => [
            'constant' => ['5.0', '7.0'],
            'namespace' => ['5.0', '8.0'],
            'class' => '5.0',
            'function' => '5.0',
            'method' => ['5.0', '7.0'],
        ],
        'finally' => [
            'constant' => ['5.5', '7.0'],
            'namespace' => ['5.5', '8.0'],
            'class' => '5.5',
            'function' => '5.5',
            'method' => ['5.5', '7.0'],
        ],
        'for' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'foreach' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'function' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'global' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'goto' => [
            'constant' => ['5.3', '7.0'],
            'namespace' => ['5.3', '8.0'],
            'class' => '5.3',
            'function' => '5.3',
            'method' => ['5.3', '7.0'],
        ],
        'if' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'implements' => [
            'constant' => ['5.0', '7.0'],
            'namespace' => ['5.0', '8.0'],
            'class' => '5.0',
            'function' => '5.0',
            'method' => ['5.0', '7.0'],
        ],
        'include' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'include_once' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'instanceof' => [
            'constant' => ['5.0', '7.0'],
            'namespace' => ['5.0', '8.0'],
            'class' => '5.0',
            'function' => '5.0',
            'method' => ['5.0', '7.0'],
        ],
        'insteadof' => [
            'constant' => ['5.4', '7.0'],
            'namespace' => ['5.4', '8.0'],
            'class' => '5.4',
            'function' => '5.4',
            'method' => ['5.4', '7.0'],
        ],
        'interface' => [
            'constant' => ['5.0', '7.0'],
            'namespace' => ['5.0', '8.0'],
            'class' => '5.0',
            'function' => '5.0',
            'method' => ['5.0', '7.0'],
        ],
        'isset' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'list' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'match' => [
            'constant' => false,
            'namespace' => false,
            'class' => '8.0',
            'function' => '8.0',
            'method' => false,
        ],
        'namespace' => [
            'constant' => ['5.3', '7.0'],
            'namespace' => '5.3',
            'class' => '5.3',
            'function' => '5.3',
            'method' => ['5.3', '7.0'],
        ],
        'new' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'or' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'print' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'private' => [
            'constant' => ['5.0', '7.0'],
            'namespace' => ['5.0', '8.0'],
            'class' => '5.0',
            'function' => '5.0',
            'method' => ['5.0', '7.0'],
        ],
        'protected' => [
            'constant' => ['5.0', '7.0'],
            'namespace' => ['5.0', '8.0'],
            'class' => '5.0',
            'function' => '5.0',
            'method' => ['5.0', '7.0'],
        ],
        'public' => [
            'constant' => ['5.0', '7.0'],
            'namespace' => ['5.0', '8.0'],
            'class' => '5.0',
            'function' => '5.0',
            'method' => ['5.0', '7.0'],
        ],
        'require' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'require_once' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'return' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'static' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'switch' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'throw' => [
            'constant' => ['5.0', '7.0'],
            'namespace' => ['5.0', '8.0'],
            'class' => '5.0',
            'function' => '5.0',
            'method' => ['5.0', '7.0'],
        ],
        'trait' => [
            'constant' => ['5.4', '7.0'],
            'namespace' => ['5.4', '8.0'],
            'class' => '5.4',
            'function' => '5.4',
            'method' => ['5.4', '7.0'],
        ],
        'try' => [
            'constant' => ['5.0', '7.0'],
            'namespace' => ['5.0', '8.0'],
            'class' => '5.0',
            'function' => '5.0',
            'method' => ['5.0', '7.0'],
        ],
        'unset' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'use' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'var' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'while' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'xor' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        'yield' => [
            'constant' => ['5.5', '7.0'],
            'namespace' => ['5.5', '8.0'],
            'class' => '5.5',
            'function' => '5.5',
            'method' => ['5.5', '7.0'],
        ],
        '__class__' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        '__dir__' => [
            'constant' => ['5.3', '7.0'],
            'namespace' => ['5.3', '8.0'],
            'class' => '5.3',
            'function' => '5.3',
            'method' => ['5.3', '7.0'],
        ],
        '__file__' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        '__function__' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        '__line__' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        '__method__' => [
            'constant' => ['4.0', '7.0'],
            'namespace' => ['4.0', '8.0'],
            'class' => '4.0',
            'function' => '4.0',
            'method' => ['4.0', '7.0'],
        ],
        '__namespace__' => [
            'constant' => ['5.3', '7.0'],
            'namespace' => ['5.3', '8.0'],
            'class' => '5.3',
            'function' => '5.3',
            'method' => ['5.3', '7.0'],
        ],
        '__trait__' => [
            'constant' => ['5.4', '7.0'],
            'namespace' => ['5.4', '8.0'],
            'class' => '5.4',
            'function' => '5.4',
            'method' => ['5.4', '7.0'],
        ],
        'int' => [
            'constant' => false,
            'namespace' => ['7.0', '8.0'],
            'class' => '7.0',
            'function' => false,
            'method' => false,
        ],
        'float' => [
            'constant' => false,
            'namespace' => ['7.0', '8.0'],
            'class' => '7.0',
            'function' => false,
            'method' => false,
        ],
        'bool' => [
            'constant' => false,
            'namespace' => ['7.0', '8.0'],
            'class' => '7.0',
            'function' => false,
            'method' => false,
        ],
        'string' => [
            'constant' => false,
            'namespace' => ['7.0', '8.0'],
            'class' => '7.0',
            'function' => false,
            'method' => false,
        ],
        'true' => [
            'constant' => false,
            'namespace' => ['7.0', '8.0'],
            'class' => '7.0',
            'function' => false,
            'method' => false,
        ],
        'false' => [
            'constant' => false,
            'namespace' => ['7.0', '8.0'],
            'class' => '7.0',
            'function' => false,
            'method' => false,
        ],
        'null' => [
            'constant' => false,
            'namespace' => ['7.0', '8.0'],
            'class' => '7.0',
            'function' => false,
            'method' => false,
        ],
        'void' => [
            'constant' => false,
            'namespace' => ['7.0', '8.0'],
            'class' => '7.0',
            'function' => false,
            'method' => false,
        ],
        'iterable' => [
            'constant' => false,
            'namespace' => ['7.1', '8.0'],
            'class' => '7.1',
            'function' => false,
            'method' => false,
        ],
        'object' => [
            'constant' => false,
            'namespace' => ['7.2', '8.0'],
            'class' => '7.2',
            'function' => false,
            'method' => false,
        ],
        'resource' => [
            'constant' => false,
            'namespace' => false,
            'class' => false,
            'function' => false,
            'method' => false,
        ],
        'mixed' => [
            'constant' => false,
            'namespace' => false,
            'class' => '8.0',
            'function' => false,
            'method' => false,
        ],
        'numeric' => [
            'constant' => false,
            'namespace' => false,
            'class' => false,
            'function' => false,
            'method' => false,
        ],
        'readonly' => [
            'constant' => false,
            'namespace' => false,
            'class' => '8.1',
            'function' => false,
            'method' => false,
        ],
    ];
}
